packages/xstate-test/src/dedupPathPlans.ts

```diff
@@ -0,0 +1,97 @@
+import { StatePath, StatePlan } from '@xstate/graph';
+import { EventObject } from 'xstate';
+import { PlanGenerator } from './types';
+
+/**
+ * Deduplicates your path plans so that A -> B
+ * is not executed separately to A -> B -> C
+ */
+export const planGeneratorWithDedup = <TState, TEvent extends EventObject>(
+  planGenerator: PlanGenerator<TState, TEvent>
+): PlanGenerator<TState, TEvent> => (behavior, options) => {
+  const pathPlans = planGenerator(behavior, options);
+
+  /**
+   * Put all plans on the same level so we can dedup them
+   */
+  const allPathsWithPlan: Array<{
+    path: StatePath<TState, TEvent>;
+    planIndex: number;
+    eventSequence: string[];
+  }> = [];
+
+  pathPlans.forEach((plan, index) => {
+    plan.paths.forEach((path) => {
+      allPathsWithPlan.push({
+        path,
+        planIndex: index,
+        eventSequence: path.steps.map((step) =>
+          options.serializeEvent(step.event)
+        )
+      });
+    });
+  });
+
+  // Sort by path length, descending
+  allPathsWithPlan.sort((a, z) => z.path.steps.length - a.path.steps.length);
+
+  const superpathsWithPlan: typeof allPathsWithPlan = [];
+
+  /**
+   * Filter out the paths that are subpaths of superpaths
+   */
+  pathLoop: for (const pathWithPlan of allPathsWithPlan) {
```

Comment:
Second time ever that I see a literal label combined with break/continue in JS. And the second time that I see this in XState - neat 😉

Reply:
That's how you know we do ✨ algorithms ✨ here